### PR TITLE
2214 navigation rework

### LINKED ISSUE #2214: Navigation Bar gets to crowded on 13" Macbook Pro
**Current Behavior**
The navigation bar items get the whole navigation very crowded and it's almost overflowing on my 13" Macbook Pro. When I resize the window, it breaks the searchbar + settings button completely.

**Expected behavior/code**
The navigation items could be resized on smaller resolutions or could even collapse automatically.

**Environment**
- Browser Version: Edge Beta 78
- OS: macOS 10.14.5
- Resolution: 2560 x 1600
- Screen size: 13"

**Possible Solution**
We could move the overflowing items into a dropdown menu similar to this: https://codepen.io/kvana/full/ZbENqr

**Additional context/Screenshots**

Default fullscreen view:
image

Resized view:
image

### PR DESCRIPTION
**Pull request by bdbch**
closes #2214 

With this PR I'm starting to rebuild the navigation/headerbar component and introduced some package updates so we can move onwards with functionalities and get around dirty ways like using a Route for the app to have access to the router.

Updated packages:
- React => 16.10.2
- ReactDOM => 16.10.2
- ReactRouterDOM => 5.1.2

I also started implement @material-ui/core since the old material-core npm package was deprecated a long time ago so we should start to reimplement components with the new library from now on + rewrite old components. What do you think about this @howardchung ?

I also implemented the `use-react-router` hook to get direct router access from any functional component.

Screenshots:

Menu default:
image

Menu open:
image

### CODE REVIEW

**Comment by bdbch** [comment]
I also reworked the SettingsGroup component with the new dropdown component:

image

image

So both work the same way

**Comment by bdbch** [comment]
I also updated redux and react-redux and fixed migration issues so we can now use react-redux hooks like `useSelector` or `useDispatch` instead of `Connect`

**Comment by howardchung** [comment]
Old material-ui is supposed to be somewhat difficult to migrate due to incompatibilities. I'd actually lean toward using a new library if we want to move old components off the deprecated library.

File: src/components/Header/Header.jsx [hunk 1]
```diff
@@ -73,6 +81,64 @@ const ToolbarHeader = styled(Toolbar)`
   }
 `;
 
+const LinkGroup = ({ navbarPages }) => {
+  const router = useReactRouter();
```
Comment (howardchung) [inline src/components/Header/Header.jsx]: I believe this is the first React Hook we're using. I don't have a ton of experience with them, do they provide a significant benefit to developer experience?

Comment (102) [reply src/components/Header/Header.jsx]: Hooks are quite convenient to use & are also awesome to be used with type analysis tools like typescript/flow

**Comment by bdbch** [comment]
@howardchung awesome!

> Old material-ui is supposed to be somewhat difficult to migrate due to incompatibilities. I'd actually lean toward using a new library if we want to move old components off the deprecated library.

Yes, it seems like they rewrote the whole thing and moved things into different packages. We could look for old rewriteable components sporadically (like for example this navigation in this PR) and update them with the new ui library.

> I believe this is the first React Hook we're using. I don't have a ton of experience with them, do they provide a significant benefit to developer experience?

Hooks are awesome when it comes to reusing logic. While in class components your logic is a part of your class or some util function, a hook can bring logic with itself and handle states, update functions etc. itself.

I've wrote some hook libs myself which you could take a look at to see what they're meant to do:

- https://github.com/bdbch/react-localstorage
- https://github.com/bdbch/react-github
- https://github.com/bdbch/react-devto

Also what this allows us to do is to stop transfering router props and the mapping of redux props with new hooks for example

```js
// Push to a specific route
const router = useReactRouter();
router.history.push('/Route');
```

or

```js
// Get strings from our redux store
const strings = useSelector(state => state.app.strings);
```

or

```js
// dispatch redux actions
const dispatch = useDispatch();
dispatch(setSearchQuery(query));
```

---

in general we can replace class components with functional components and extract logic out of them into separated hooks or functions. If you're interested, I could add the next functionality as a FC and ping you in the PR so you can take a look. :)

**Pull request merged**